server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,275 @@
+                  ? null
+                  : Arrays.asList(key, v.iterator().next()));
+   }
+
+   @Listener(clustered = true)
+   public static class PubSubListener {
+      private final AdvancedCache<byte[], Object> cache;
+      private volatile ScheduledFuture<?> scheduledTimer;
+      private final Resp3Handler handler;
+      private final PollListenerSynchronizer synchronizer;
+
+      private PubSubListener(List<byte[]> filterKeys, Resp3Handler handler, AdvancedCache<byte[], Object> cache,
+            EmbeddedMultimapListCache<byte[], byte[]> mml) {
+         this.cache = cache;
+         this.handler = handler;
+         this.synchronizer = new PollListenerSynchronizer(filterKeys, mml);
+
+         synchronizer.resultFuture.whenComplete((ignore_v, ignore_t) -> {
+            deleteTimer();
+            cache.removeListenerAsync(this);
+         });
+      }
+
+      public CompletableFuture<Collection<byte[]>> getFuture() {
+         return synchronizer.resultFuture;
+      }
+
+      private void startTimer(long timeout) {
+         deleteTimer();
+         scheduledTimer = (timeout > 0) ? handler.getScheduler().schedule(() -> {
+            cache.removeListenerAsync(this);
+            synchronizer.complete(null);
+         }, timeout, TimeUnit.MILLISECONDS) : null;
+      }
+
+      private void deleteTimer() {
+         if (scheduledTimer != null)
+            scheduledTimer.cancel(true);
+         scheduledTimer = null;
+      }
+
+      @CacheEntryCreated
+      @CacheEntryModified
+      public void onEvent(CacheEntryEvent<Object, Object> entryEvent) {
+         try {
+            if (entryEvent.getValue() instanceof ListBucket) {
+               byte[] key = unwrapKey(entryEvent.getKey());
+               synchronizer.onEvent(key);
+            }
+         } catch (Exception ex) {
+            synchronizer.completeExceptionally(ex);
+         }
+      }
+
+      private byte[] unwrapKey(Object key) {
+         return key instanceof WrappedByteArray
+               ? ((WrappedByteArray) key).getBytes()
+               : (byte[]) key;
+      }
+   }
+
+   /**
+    * PollListenerSynchronizer
+    *
+    * This class synchronizes the access to a CompletableFuture `resultFuture` so
+    * that its final value will be completed either
+    * - with value v by an onListenerAdded() call if a not null value is found
+    * - otherwise by onEvent(k) call if the referred entry is not null or empty
+    *
+    */
+   public static class PollListenerSynchronizer {
+      private final ConcurrentLinkedQueue<byte[]> keyQueue;
+      private final AtomicInteger eventsCounter;
+      private final CompletableFuture<Collection<byte[]>> resultFuture;
+      private final EmbeddedMultimapListCache<byte[], byte[]> multimapList;
+      private final List<byte[]> keys;
+      private final BiConsumer<? super Collection<byte[]>, ? super Throwable> whenCompleteConsumer;
+      private final AtomicBoolean canPollJustEventKey;
```

Comment:
Since we're talking about the two variables here (status, keyQueue), I think there is no way to keep them in sync without the synchronized blocks. It would be nice to have everything with just the Atomic* variables, though.

Reply:
current approach is actually based only on `eventCounter`. 
`canPollJustEventKey` is an optimization to avoid a full poll after the first poll is complete, if I'm corrent an out of sync here should result running a full poll more the necessary.
Will suggested this approach using an atomic counter to match requests and executions, imo it sounds more robust that using states